go/libraries/doltcore/table/untyped/csv/reader.go

```diff
@@ -153,32 +204,200 @@ func (csvr *CSVReader) Close(ctx context.Context) error {
 	}
 }
 
-func (csvr *CSVReader) parseRow(line string) (row.Row, error) {
-	colVals, err := csvSplitLine(line, csvr.info.Delim, csvr.info.EscapeQuotes)
+// Functions below this line are borrowed or adapted from encoding/csv/reader.go
 
-	if err != nil {
-		return nil, table.NewBadRow(nil, err.Error())
+func validDelim(s string) bool {
+	return !(strings.Contains(s, "\"") ||
+		strings.Contains(s, "\r") ||
+		strings.Contains(s, "\n") ||
+		strings.Contains(s, string(0xFFFD))) // Unicode replacement char
+}
+
+// nextRune returns the next rune in b or utf8.RuneError.
+func nextRune(b []byte) rune {
+	r, _ := utf8.DecodeRune(b)
+	return r
+}
+
+func atDelim(line []byte, delim []byte) bool {
+	if len(line) < len(delim) {
+		return false
+	}
+	return bytes.Compare(line[:len(delim)], delim) == 0
+}
+
+func lengthNL(b []byte) int {
+	if len(b) > 0 && b[len(b)-1] == '\n' {
+		return 1
 	}
+	return 0
+}
 
-	sch := csvr.sch
-	allCols := sch.GetAllCols()
-	numCols := allCols.Size()
-	if len(colVals) != numCols {
-		return nil, table.NewBadRow(nil,
-			fmt.Sprintf("csv reader's schema expects %d fields, but line only has %d values.", numCols, len(colVals)),
-			fmt.Sprintf("line: '%s'", line),
-		)
+// readLine reads the next line (with the trailing endline).
+// If EOF is hit without a trailing endline, it will be omitted.
+// If some bytes were read, then the error is never io.EOF.
+// The result is only valid until the next call to readLine.
+func (csvr *CSVReader) readLine() ([]byte, error) {
+	line, err := csvr.bRd.ReadSlice('\n')
+	if err == bufio.ErrBufferFull {
+		csvr.rawBuffer = append(csvr.rawBuffer[:0], line...)
+		for err == bufio.ErrBufferFull {
+			line, err = csvr.bRd.ReadSlice('\n')
+			csvr.rawBuffer = append(csvr.rawBuffer, line...)
+		}
+		line = csvr.rawBuffer
+	}
+	if len(line) > 0 && err == io.EOF {
+		err = nil
+		// For backwards compatibility, drop trailing \r before EOF.
+		if line[len(line)-1] == '\r' {
+			line = line[:len(line)-1]
+		}
+	}
+	csvr.numLine++
+	// Normalize \r\n to \n on all input lines.
+	if n := len(line); n >= 2 && line[n-2] == '\r' && line[n-1] == '\n' {
+		line[n-2] = '\n'
+		line = line[:n-1]
 	}
+	return line, err
+}
 
-	taggedVals := make(row.TaggedValues)
-	for i := 0; i < allCols.Size(); i++ {
-		col := allCols.GetByIndex(i)
-		if colVals[i] == nil {
-			taggedVals[col.Tag] = nil
-			continue
+func (csvr *CSVReader) csvReadRecords(dst []*string) ([]*string, error) {
```

Comment:
this function is a monster. Would be nice if you could break some of the chunks up into functions with self documenting names.